Accept pre-tokenized references & hypothesis for METEOR calculation

yutanakamura-tky opened the pull request:
I added an option to input pre-tokenized sentence(s) for METEOR calculation.

At present, references & hypothesis are always tokenized with `str.split()` (Lines 30-31 of `meteor_score.py`):

```
    hypothesis_list = list(enumerate(preprocess(hypothesis).split()))
    reference_list = list(enumerate(preprocess(reference).split()))
```

However, we have trouble calculating METEOR in a language that does not separate words with spaces (e.g., Japanese).

In this pull request, I have changed `_generate_enums` to accept pre-tokenized reference & hypothesis in addition to untokenized reference & hypothesis so that we can use a customized tokenizer.

Other functions have been changed accordingly.

yutanakamura-tky committed:
accept pre-tokenized references & hypothesis

tomaarsen commented:
@yutanakamura-tky Heya! We've started using https://pre-commit.com/ pretty recently. With this program, some tests are ran before it lets you commit. With [our configuration](https://github.com/nltk/nltk/blob/develop/.pre-commit-config.yaml), it means that some commands are ran to ensure some code quality. In particular, the `black` formatter. As you can see in the [test output](https://github.com/nltk/nltk/pull/2822/checks?check_run_id=3722726809), the `black` formatter complained that the changes you are trying to push are not formatted according to the `black` formatting rules.

My recommendation is to run `pre-commit install` (`pre-commit` should have been `pip install`-ed already, but otherwise you can use `pip install pre-commit`). This will install the pre-commit hooks, causing it to run all of those hooks (e.g. formatters, verifiers, etc.) when you make another commit. If you then try to push another commit, the `black` hook will reformat your changes, and you can push the correctly formatted changes. This might seem a little bit confusing, but it's more intuitive in practice. Oh, also, `pre-commit` only affects this repository, you don't have to worry about it affecting other projects.

Upon installing, you may run `pre-commit run --all`, which applies formatting and such on all files, and in particular on your changes. You may push these changes, so the Continuous Integration stops complaining.

Let me know if you have any issues.

yutanakamura-tky committed:
accept pre-tokenized references & hypothesis

yutanakamura-tky committed:
applied code formatter

yutanakamura-tky commented:
@tomaarsen 
Thank you! I have updated this pull request using `pre-commit`.

tomaarsen committed:
wn.NOUN -> 'n' (#2823)

tomaarsen committed:
Resolved undefined variable bug in raising Exception for RegexpTagger (#2821)

stevenbird commented:
I would prefer to require tokenised input (a breaking change) than to have `str or list(str)` parameter

yutanakamura-tky committed:
breaking changes: always require pre-tokenized sentence(s) for METEOR calculation

yutanakamura-tky committed:
modified test

yutanakamura-tky committed:
fixed typo & deleted unneeded "preprocess"

yutanakamura-tky committed:
deleted unneeded "preprocess" further

yutanakamura-tky commented:
@stevenbird 
Thank you! I have made the following changes:

#### 1. Enforcement of tokenized inputs

I have updated the function implementations, docstrings, and examples.

#### 2. Modification of the test code
I have deleted the following code in `nltk/test/unit/translate/test_meteor.py` because it assumed that the input was given not in `list(str)` but `str` (Lines 12-13):

```
        # no `preprocess` argument
        self.assertRaises(TypeError, meteor_score, reference, candidate)
```

I have also changed the code as follows to examine better if the preprocessing works correctly:
```
        reference = [["this", "is", "a", "test"], ["this", "is" "test"]]
        candidate = ["THIS", "Is", "a", "tEST"]

        score = meteor_score(reference, candidate, preprocess=str.lower)
```

I have confirmed locally that it passes `pytest nltk/test/unit/translate/test_meteor.py`.

yutanakamura-tky commented:
The force-push from 5db9c72 to ea66229 was made to amend the commit message of 3d0d272 from "accept pre-tokenized references & hypothesis" to "applied code formatter."
Sorry for any inconvenience.

yutanakamura-tky committed:
type check & error message of _generate_enums()

tomaarsen commented on nltk/translate/meteor_score.py:
```diff
@@ -38,10 +49,10 @@ def exact_match(hypothesis, reference):
+    :param hypothesis: pre-tokenized hypothesis
+    :type hypothesis: array-like(str)
+    :param reference: pre-tokenized reference
+    :type hypothesis: array-like(str)
```
```suggestion
    :type reference: array-like(str)
```
I believe this is a small copy-paste error. No biggie.

tomaarsen commented on nltk/translate/meteor_score.py:
```diff
@@ -111,10 +122,10 @@ def stem_match(hypothesis, reference, stemmer=PorterStemmer()):
+    :param hypothesis: pre-tokenized hypothesis
+    :type hypothesis: array-like(str)
+    :param reference: pre-tokenized reference
+    :type hypothesis: array-like(str)
```
```suggestion
    :type reference: array-like(str)
```
idem.

tomaarsen commented on nltk/translate/meteor_score.py:
```diff
@@ -296,13 +316,13 @@ def single_meteor_score(
-    :param hypothesis: a hypothesis sentence
-    :type hypothesis: str
+    :param reference: pre-tokenized reference
+    :type hypothesis: array-like(str)
```
```suggestion
    :type reference: array-like(str)
```
idem.

tomaarsen commented on nltk/translate/meteor_score.py:
```diff
@@ -376,13 +396,13 @@ def meteor_score(
-    :param hypothesis: a hypothesis sentence
-    :type hypothesis: str
+    :param references: pre-tokenized reference sentences
+    :type hypothesis: 2D array-like(str)
```
```suggestion
    :type references: 2D array-like(str)
```
idem.

tomaarsen commented on nltk/translate/meteor_score.py:
```diff
@@ -8,27 +8,38 @@
 
 
 from itertools import chain, product
+from typing import Iterable
```
Does this get used? I'm personally down to use the Python 3.5+ typing in signatures, and use e.g.
```python
def _generate_enums(
    hypothesis: Iterable[str],
    reference: Iterable[str],
    preprocess: Callable[[str], str] = str.lower
) -> Tuple[List[Tuple[str, str]], List[Tuple[str, str]]]:
    """...
    
    :param hypothesis: pre-tokenized hypothesis
    :type hypothesis: Iterable[str]
    :param reference: pre-tokenized reference
    :type reference: Iterable[str]
    :preprocess: preprocessing method (default str.lower)
    :type preprocess: Callable[[str], str]
    :return: enumerated words list
    :rtype: list of 2D tuples, list of 2D tuples
    :rtype: Tuple[List[Tuple[str, str]], List[Tuple[str, str]]]
    """
    ...
```
*However*, this isn't necessarily in the scope of your PR. This is a change I'd like to see throughout the entire codebase, so you can just leave it like it was if you prefer. At that point it might also be overkill to specify the type *both* in the signature and in the docstring.

tomaarsen commented:
> I would prefer to require tokenised input (a breaking change) than to have `str or list(str)` parameter

I believe that all the other translation scores, i.e. BLEU, CHRF, GLEU, NIST and RIBES also use pre-tokenized inputs (Correct me if I'm wrong). As a result, I'm all in favor of modifying the METEOR score so that it also takes tokenised input.

tomaarsen committed:
Merge branch 'develop' of https://github.com/nltk/nltk into pr/2822

tomaarsen committed:
Applied tokenization on hypotheses and references in meteor.doctest

tomaarsen committed:
Applied tokenization on hypotheses and references in meteor.doctest

tomaarsen commented:
I've applied tokenization on the hypotheses and references in `meteor.doctest`, so the tests use tokenized inputs rather than the original string ones. The tests pass now! 🎉

yutanakamura-tky committed:
added type hints & modified local variable names

yutanakamura-tky committed:
Merge branch 'modified_meteor' of github.com:yutanakamura-tky/nltk into modified_meteor

yutanakamura-tky commented:
@tomaarsen 
Thank you for your review comments & updating `meteor.doctest`!
I have made some changes as follows:

### 1. Addition of Type Hints
I have added type hints to all methods in `meteor_score.py`.
Accordingly, I have removed type descriptions in the docstrings.

### 2. Minor Improvement of Local Variable Names
I have changed some local variable names to make their roles clearer.

- `_generate_enums()`
  - `hypothesis_list` -> `enum_hypothesis_list` 
  - `reference_list` -> `enum_reference_list` 
- `exact_match()`
  - `hypothesis_list` -> `enum_hypothesis_list` 
  - `reference_list` -> `enum_reference_list` 
- `_enum_stem_match()`
  - `stemmed_enum_list1` -> `stemmed_enum_hypothesis_list` 
  - `stemmed_enum_list2` -> `stemmed_enum_reference_list`

yutanakamura-tky committed:
more detailed type hints for Callable

tomaarsen committed:
Placed .pop's on two lines

I'm guessing this was leftover from when it was wrapped with print() for debugging?

tomaarsen committed:
Replaced last instance of array-like(...) with Iterable[...]

tomaarsen reviewed (approved):
Looks good to me! I'm a great fan of the modernisation, and the modified functionality is wonderful too, it makes the METEOR score calculation similar to the other translation score metrics.

yutanakamura-tky commented:
@tomaarsen 
Thank you so much!
I want just one more piece of advice about variable typing.

The ongoing docstring in the `develop` branch says that `meteor_score()` takes `float` for the variable `beta`, while it defaults to 3:

```
def meteor_score(
    references: Iterable[Iterable[str]],
    hypothesis: Iterable[str],
    preprocess: Callable[[str], str] = str.lower,
    stemmer: StemmerI = PorterStemmer(),
    wordnet: WordNetCorpusReader = wordnet,
    alpha: float = 0.9,
    beta: float = 3,
    gamma: float = 0.5,
)
```

The "3" derives from the power of penalty in the original METEOR article.
I am wondering which is better, `beta: Union[int, float]=3` or `beta: float=3.0`.

tomaarsen commented:
@yutanakamura-tky I think the best solution in that situation is simply to change `beta: float = 3` into `beta: float = 3.0`. That way it's extra clear that this value can be/is meant to be a float. That said, users are free to use e.g. `4`, because Python automatically converts ints to floats whenever requires anyways. 
Good catch, by the way!

tomaarsen committed:
Replaced default value for beta in meteor_score from 3 to 3.0

yutanakamura-tky committed:
Replaced default value for beta in meteor_score from 3 to 3.0 also in single_meteor_score()

yutanakamura-tky commented:
@tomaarsen 
Thank you so much! I have also made the same change to `single_meteor_score()`.
There are no further changes that I am planning to make.

tomaarsen commented:
Wonderful, it all looks great to me. I'll leave the PR like this for now, so perhaps another team member can have a look at it too.

dannysepler commented:
nice! i love the addition of type hints anywhere in NLTK

~~given that this is a breaking change, should we have a changelog entry? after that, lgtm!~~ (EDIT: Realizing now that only Steven has edited that file in recent history. Sorry!)

stevenbird commented:
Thank you @yutanakamura-tky, @tomaarsen, @dannysepler.

Witiko commented:
@stevenbird Please, consider incrementing at least the minor version the next time you introduce a breaking change.

stevenbird commented:
@Witiko: noted, my apologies
